lms/envs/production.py

```diff
@@ -490,6 +490,18 @@ def get_env_setting(setting):
 # Get the MODULESTORE from auth.json, but if it doesn't exist,
 # use the one from common.py
 MODULESTORE = convert_module_store_setting_if_needed(AUTH_TOKENS.get('MODULESTORE', MODULESTORE))
+
+# After conversion above, the modulestore will have a "stores" list with all defined stores, for all stores, add the
+# fs_root entry to derived collection so that if it's a callable it can be resolved.  We need to do this because the
+# `derived_collection_entry` takes an exact index value but the config file might have overidden the number of stores
+# and so we can't be sure that the 2 we define in common.py will be there when we try to derive settings.  This could
+# lead to execptions being thrown when the `derive_settings` call later in this file tries to update settings.  We call
+# the derived_collection_entry function here to ensure that we update the fs_root for any callables that remain after
+# we've updated the MODULESTORE setting from our config file.
+for idx, store in enumerate(MODULESTORE['default']['OPTIONS']['stores']):
+    if 'OPTIONS' in store and 'fs_root' in store["OPTIONS"]:
+        derived_collection_entry('MODULESTORE', 'default', 'OPTIONS', 'stores', idx, 'OPTIONS', 'fs_root')
+
```

Comment:
@feanil this looks like it works, and the comment is very helpful, but I am reluctant to add such specific logic to production.py if there's any simpler way to do it. Were you able to try my suggestion?

> @feanil maybe you could you work around this by adding MODULESTORE to [KEYS_WITH_MERGED_VALUES](https://github.com/openedx/edx-platform/blob/0544c067058e3ec5dd22af695aeb682494a49e0c/lms/envs/production.py#L77), and then overriding MODULESTORE (but just the entry you need) in minimal.yml? Haven't tried it myself, but my understanding is that this is the case that KEYS_WITH_MERGED_VALUES exists for.

Reply:
@kdmccormick yes, I looked at that, unfortunately, that's not quite how that variable works.  By adding it to that list, it essentially tells production.py to not automatically update `MODULESTORE` from anything it sees in the config yaml file.  We would then have to add logic in `production.py` to manually load or partially load the modulestore settings.  Given that there are people who already override the whole thing.  We would need to be able to manually suport that if we were to add modulestore to `KEYS_WITH_MERGED_VALUES` which would essentially put us back where we are now.

I agree that this is very specific logic but at least it's explicit and clear and can be targeted for removal later by changing how we structure loading modulestore settings. (e.g. We force everyone to use Mixedmodulestore anyway, why do we need that to be part of the settings structure, it made sense for transitioning but we should have completed the transition by removing this oddity and simplifying what info operators need to provide)

Reply:
Sounds good, thanks for checking it out. 

> We force everyone to use Mixedmodulestore anyway, why do we need that to be part of the settings structure, it made sense for transitioning but we should have completed the transition by removing this oddity and simplifying what info operators need to provide

+100